pkg/ldbwriter/ldb_callback_writer.go

```diff
@@ -3,33 +3,105 @@ package ldbwriter
 import (
 	"context"
 	"database/sql"
-
 	"github.com/segmentio/ctlstore/pkg/schema"
 	"github.com/segmentio/ctlstore/pkg/sqlite"
 	"github.com/segmentio/events/v2"
+	"github.com/segmentio/stats/v4"
 )
 
 // CallbackWriter is an LDBWriter that delegates to another
 // writer and then, upon a successful write, executes N callbacks.
 type CallbackWriter struct {
-	DB           *sql.DB
-	Delegate     LDBWriter
-	Callbacks    []LDBWriteCallback
+	DB        *sql.DB
+	Delegate  LDBWriter
+	Callbacks []LDBWriteCallback
+	// Buffer between SQLite preupdate Hook and this code
 	ChangeBuffer *sqlite.SQLChangeBuffer
+	// Accumulated changes across multiple ApplyDMLStatement calls
+	transactionChanges []sqlite.SQLiteWatchChange
```

Comment:
should we consider a cap on the size of a transaction so we don't buffer too much?

Reply:
Interesting proposal, issues around that are where my head was at as I was adding the metrics.

This consideration relates to the "support SoR transactions" project.  We don't yet know how large SoR transactions can get, especially for those that we care about for that project. Putting some cap here with a loud complaint that we exceeded it (log + metric) and then dumping the currently buffered changes to the changelog (i.e., invoking callbacks) seems a fine behavior for now.

We know that the changes would not be greater than 200 right now based on hardcoded limits elsewhere & the behavior of the `RELACE INTO`, so perhaps just setting it to like 500 for now would be good. i.e.,
* `REPLACE INTO` is translated by SQLite into a `DELETE` op then an `INSERT` op ➡️ 2 changes.
* max of 100 entries in a ledger transaction: https://github.com/segmentio/ctlstore/blob/365206507ae8ef2f4a70bf416487a22bb2b0803a/pkg/executive/db_executive.go#L455-L458

Reply:
that sounds good to me

Reply:
This comment is not relevant to today's transactions since they are size limited but rather to the future when we are supporting SoR transactions. A sane number of statements within a transaction is smart, but by doing this we need to ensure this limit is well understood externally. Once we say we support transactions we need to be sure we really do.